proposals/transcoding_invalid_identifiers/README.md

```diff
@@ -0,0 +1,353 @@
+|               |   |     |                               |     |      |
+| V<sub>j</sub> |   |     | value = V<sub>j</sub> * N + i |     |      |
+|               |   |     |                               |     |      |
+
+Thus, the extended code (V<sub>j</sub>) and the position (i) can be extracted as (_floor(value / N)_) and (_value % N_) 
+respectively.
+
+## Decoding procedure
+
+`TfMakeValidIdentifier` had no reverse function, it was not possible. Since our function is bijective we can create 
+a decoding mechanism for the proposed solution.
+
+The decoding procedure follows the reverse process:
+- Remove prefix.
+- Copy the basic codes.
+- For the extended codes (i.e. encoding section):
+  - Let `value` be the variable length integer read.
+  - Increase code value `code` by `value / N` (due to delta encoding).
+  - Let position `pos` be `value % N`.
+  - Insert code value `code` at position `pos`.
+  - Increase N.
+
+It is important to notice that whereas the proposed encoding could generate different values depending on what 
+character set is considered (i.e. ASCII vs UTF-8 XID), decoding is agnostic to the character set and will always result
+into the original string.
+
+## Differences to Punycode
+
+A summary of the differences against Punycode are:
+- The separating character changes from `-` to `_`. Since `-` is invalid character in OpenUSD.
+- Delta encoding starts from `0` instead of `128`. This is to account the fact not all ASCII's are allowed in OpenUSD.
+- The base representation changes from `36` in Punycode to `62` in this implementation (to represent more information
+in less characters).
+- Threshold is constant in this implementation. There is no loss of performance or memory representation, since our 
+`base` is also increased, and simplifies implementation.
+
+# Examples
+
+The above example:
+
+| Original        | Transcoding           |
+|-----------------|-----------------------|
+| 012-345-678/9.0 | tn__01234567890_lG7QQ |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("012-345-678/9.0") == "tn__01234567890_lG7QQ"
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("tn__01234567890_lG7QQ") == "012-345-678/9.0"
+);
+```
+
+Encoding valid identifiers produces no changes.
+
+| Original      | Transcoding   |
+|---------------|---------------|
+| id12345_abcde | id12345_abcde |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("id12345_abcde") == "id12345_abcde"
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("id12345_abcde") == "id12345_abcde"
+);
+```
+
+An encoded identifier is already a valid identifier, and it will result in itself.
+
+| Original              | Transcoding           |
+|-----------------------|-----------------------|
+| tn__01234567890_lG7QQ | tn__01234567890_lG7QQ |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("tn__01234567890_lG7QQ") == "tn__01234567890_lG7QQ"
+);
+```
+
+Existing valid identifiers with `tn__` prefix will produce no changes.
```

Comment:
At one point we had discussed that the right behavior is to attempt to decode and then reencode.  Motivation--  Let's say I have `"tn__MünchenGermany_rEi5`, an identifier previously encoded with `SdfBoostringEncodeIdentifier`, and I want to ensure it's ASCII encode with `SdfBoostringEncodeAsciiIdentiifer`.  It also can function as a validator.  Run `SdfBootrstringEncodeIdentifier` as a way of ensuring that an identifier is properly encoded.

Reply:
Or actually vice versa-- I'm leaving a domain that required Ascii identifiers and now I want to "upgrade" to Utf8.

Reply:
I think the input for encode should by any UTF-8 string and we do not really have any interpretation of what that string is. The output of encoding should be:
- empty string: in case of invalid UTF-8
- the same string: if the string is already in the domain of valid characters
- encoded string (i.e. with `tn__` prefix): if the UTF-8 is not in the domain of valid characters.

I think the 3 proposed methods pose the minimum set of operations to fix this problem:
```cpp
std::optional<std::string> SdfBootstringReencodeIdentifier(const std::string& identifier) {
   std::string originalIdentifier = SdfBootstringDecodeIdentifier(identifier).value_or(identifier);
   return SdfBootstringEncodeIdentifier(originalIdentifier);
}
```
That can be added into the set of methods to the API, and I think the intent is clear: it will attempt to check the passed identifier is a valid encoded identifier and it will encode it again.